src/unix/getaddrinfo.c

```diff
@@ -35,6 +35,44 @@
 /* EAI_* constants. */
 #include <netdb.h>
 
+static void uv__getaddrinfo_common_done(uv_getaddrinfo_t* req, int status);
+
+#if defined(_USE_LIBINFO)
+
+typedef void (getaddrinfo_async_callback)(int32_t, struct addrinfo *, void *);
+extern int32_t (getaddrinfo_async_start)(mach_port_t *, const char *, const char *, const struct addrinfo *, getaddrinfo_async_callback, void *);
+extern int32_t (getaddrinfo_async_handle_reply)(void *);
+extern void (getaddrinfo_async_cancel)(mach_port_t);
+
+static void uv__getaddrinfo_async_done(int32_t status, struct addrinfo *addrinfo, void *context) {
+  uv_getaddrinfo_t *req;
+
+  req = context;
+  req->addrinfo = addrinfo;
+  if (status == UV_UNKNOWN || status == UV_EAI_CANCELED)
+    req->retcode = status;
+  else
+    req->retcode = uv__getaddrinfo_translate_error(status);
+  uv__io_close(req->loop, &req->async_addrinfo_io);
+  uv__req_unregister(req->loop, req);
+  uv__getaddrinfo_common_done(req, status);
+}
+
+static void uv__getaddrinfo_async_work(uv_loop_t* loop, uv__io_t* w, unsigned int fflags) {
+  READ_MACH_MSG_AND_HANDLE_DATA(w,
```

Comment:
```c
uv_getaddrinfo_t* req = container_of(w, uv_getaddrinfo_t, async_addrinfo_io);
```

should allow you to remove `w->data` asked by @bnoordhuis

Reply:
I like the idea, however I think that @bnoordhuis asked to remove the async_addrinfo_io all together, right?

Reply:
That's right.

Reply:
@cabelitos by removing the `async_addrinfo_io` and reusing the `uv__work` storage you can add the pointer in there as well (see what I called `io_data`).